presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/TypeValidator.java

```diff
@@ -154,10 +155,11 @@ private void checkFunctionCall(Map<Symbol, FunctionCall> functionCalls)
             }
         }
 
-        private static void verifyTypeSignature(Symbol symbol, TypeSignature expected, TypeSignature actual)
+        private void verifyTypeSignature(Symbol symbol, TypeSignature expected, TypeSignature actual)
         {
             // UNKNOWN should be considered as a wildcard type, which matches all the other types
-            if (!actual.equals(UNKNOWN.getTypeSignature())) {
+            TypeManager typeManager = metadata.getTypeManager();
+            if (!actual.equals(UNKNOWN.getTypeSignature()) && !typeManager.isTypeOnlyCoercion(typeManager.getType(actual), typeManager.getType(expected))) {
```

Comment:
That's part of the reason but not the biggest problem. The main issue is that it requires inserting casts all over the place, which breaks some optimizers that manipulate expressions and requires add pre-projections for some operations.

It's one of the things we need to clean up after all the decimal/varchar changes are in.